File: src/cpp/Im.cpp
#include "Im.hpp"

#include <cmath>

namespace {
const scalar_t e = 2.718281828459045235360287471352;
const scalar_t QT = std::pow(2.3, (34.0 - 21.0) / 10.0);
const scalar_t gIm_bar = 0.00001;
}

scalar_t Im_derivatives(scalar_t m, scalar_t V) {
    const scalar_t mAlpha = 3.3e-3 * std::pow(e, (2.5 * 0.04 * (V + 35)));
    const scalar_t mBeta = 3.3e-3 * std::pow(e, (-2.5 * 0.04 * (V + 35)));
    const scalar_t mInf = mAlpha / (mAlpha + mBeta);
    const scalar_t mTau = (1.0 / (mAlpha + mBeta)) / QT;

    return (mInf - m) / mTau;
}

scalar_t Im_current(scalar_t m, scalar_t V, scalar_t E_K, scalar_t dt) {
    const scalar_t mAlpha = 3.3e-3 * std::pow(e, (2.5 * 0.04 * (V + 35)));
    const scalar_t mBeta = 3.3e-3 * std::pow(e, (-2.5 * 0.04 * (V + 35)));
    const scalar_t mInf = mAlpha / (mAlpha + mBeta);
    const scalar_t mTau = (1.0 / (mAlpha + mBeta)) / QT;

    m = mInf + (m - mInf) * std::pow(e, (-dt / mTau));
    const scalar_t gIm = gIm_bar * m;

    const scalar_t I_K = gIm * (V - E_K);
    return I_K;
}

File: src/conductances/Im.py
"""
Im: M-type potasssium Conductance

Extracted from BBP NEURON mechanism: Im.mod
Neuron type: pyramidal neuron

State variables:
    m: activation gate
"""

# temperature correction
QT = 2.3**((34 - 21) / 10)

# euler's constant to 30 digits of precision
e = 2.718281828459045235360287471352

# maximal peak conductance density of channel per unit membrane area (S/cm^2)
gIm_bar = 0.00001

def derivatives(m, V):
    """
    Compute steady-state value and time constant,
    then compute time derivative of the gating variable.

    Inputs:
        - m (float): gating variable
        - V (float): membrane potential (mV)

    Returns: 
        - dm_dt (float): time derivative of m
    """

    mAlpha = 3.3e-3 * e**(2.5 * 0.04 * (V + 35))
    mBeta = 3.3e-3 * e**(-2.5 * 0.04 * (V + 35))
    mInf = mAlpha / (mAlpha + mBeta)
    mTau = (1 / (mAlpha + mBeta)) / QT

    dm_dt = (mInf - m) / mTau

    return dm_dt

def current(m, V, E_K, dt=0.025):
    """
    Compute current based on gating variables and membrane potential.

    Inputs:
        - m (float): gating variable
        - V (float): membrane potential (mV)
        - E_K (float): potassium reversal potential (mV)
        - time step (float): time step of derivative measurement

    Returns: 
        - I_K (float): current (mA/cm^2)
    """
    mAlpha = 3.3e-3 * e**(2.5 * 0.04 * (V + 35))
    mBeta = 3.3e-3 * e**(-2.5 * 0.04 * (V + 35))
    mInf = mAlpha / (mAlpha + mBeta)
    mTau = (1 / (mAlpha + mBeta)) / QT
    
    # cnexp update
    m = mInf + (m - mInf) * e**(-dt/mTau)
    gIm = gIm_bar * m
    
    I_K = gIm * (V - E_K)
    return I_K